Create decoder class

erichan1 opened the pull request:
Summary: Implement decoder-only layer, which doesn't currently exist in torch. Implement incremental and forced decoding with a new multiheadattention and decoder forward pass. Rather similar to the transformer_encoder_layer_fwd. Not a public facing API, although may become public facing eventually. 

Stacked on top of https://github.com/pytorch/pytorch/pull/79437.

Test Plan: See D36140513 numerical tests.

Reviewed By: mikekgfb

Differential Revision: D36987004

zrphercule commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -478,5 +490,182 @@ std::tuple<Tensor, Tensor> native_multi_head_attention(
+  // shape: [B, num_head, T, T]
+  // TODO: long-term, have a kernel that works with
+  // NestedTensor directly if there is no mask passed
+  qkt = masked_softmax(qkt, mask, query);
```
Do we even have nested masedk softmax already?

erichan1 replied:
I copied this from transformer_encoder_layer_fwd, so I don't believe so. On testing, would you recommend unit tests that just check for invalid shapes like so https://github.com/pytorch/pytorch/blob/543919cfc8d7014dfa365b55c459fa2b3f746891/test/test_nn.py#L5878? I don't see any numerical testing for transformer_encoder_layer_fwd. 

I have numerical testing in an internal benchmarking script where I compare against fairseq, but putting that test in torch.testing would mean I need to import fairseq in the test, which isn't a reliable dependency. I can't compare against nn.TransformerDecoder because we don't have an existing decoder-only layer, only encoder-decoder.

erichan1 replied:
Tests are in test_transformers.py here https://github.com/pytorch/pytorch/pull/79576.

zrphercule reviewed (commented):
Could you plz add some unit test in test/test_nn.py ? Seems these changes are no where tested

erichan1 committed:
Refactor out QKV in projection (#79437)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/79437

Refactor to reduce amount of copied code for decoder by finding common chunks for encoder and decoder. QKV in projection is a reasonable unit to copy out.

Test Plan:
buck run mode/opt -c fbcode.platform=platform010 -c fbcode.enable_gpu_sections=true -c fbcode.nvcc_arch=a100 //pytext/fb/tools:benchmark_transformers -- transformer --batch-size 64 --avg-sequence-length 235 --max-sequence-length 256 --iters 100  --module native

Benchmark and numerical tests work fine.

Differential Revision: D36138504

fbshipit-source-id: e7fb4b961df258e12bbd3af24fc092ac56743d46

erichan1 committed:
Create decoder class (#79438)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/79438

Implement incremental and forced decoding with a new multiheadattention and decoder forward pass. Rather similar to the transformer_encoder_layer_fwd

Test Plan: See D36140513 numerical tests.

Reviewed By: mikekgfb

Differential Revision: D36987004

fbshipit-source-id: 98fd7eaabd9a7f161dcb418ce8104e70b3cd804b

zrphercule reviewed (approved):
Great job!

erichan1 commented:
@pytorchbot merge -g
